soliket/ccl.py

```diff
@@ -109,58 +122,70 @@ def calculate(self, state, want_derived=True, **params_values_dict):
-        # Boltzmann code in general; ultimately CCL should use more physical
-        # inputs for anything of use in general models.
-        # For now just compute from CCL itself to keep it happy:
-        cosmo = ccl.Cosmology(Omega_c=Omega_c, Omega_b=Omega_b, h=h,
-                              n_s=self.provider.get_param('ns'),
-                              A_s=self.provider.get_param('As'))
         # Array z is sorted in ascending order. CCL requires an ascending scale
         # factor as input
         a = 1. / (1 + self.z[::-1])
-        growth = ccl.background.growth_factor(cosmo, a)
-        fgrowth = ccl.background.growth_rate(cosmo, a)
-        # In order to use CCL with input arrays, the cosmology object needs
-        # to be reset. This should be improved...
-        cosmo = ccl.Cosmology(Omega_c=Omega_c, Omega_b=Omega_b, h=h,
-                              n_s=self.provider.get_param('ns'),
-                              A_s=self.provider.get_param('As'), **self.extra_args)
-        cosmo._set_background_from_arrays(a_array=a,
-                                          chi_array=distance,
-                                          hoh0_array=E_of_z,
-                                          growth_array=growth,
-                                          fgrowth_array=fgrowth)
+        #growth = ccl.background.growth_factor(cosmo, a)
+        #fgrowth = ccl.background.growth_rate(cosmo, a)
+
+        # Create a CCL cosmology object. Because we are giving it background 
+        # quantities, it should not depend on the cosmology parameters given
+        cosmo = self.ccl.CosmologyCalculator(
+            Omega_c=Omega_c, Omega_b=Omega_b, h=h, sigma8=0.8, n_s=0.96,
+            background={'a': a,
+                    'chi': distance,
+                    'h_over_h0': E_of_z},
+        )
 
         if self.kmax:
             for pair in self._var_pairs:
                 # Get the matter power spectrum:
                 k, z, Pk_lin = self.provider.get_Pk_grid(var_pair=pair, nonlinear=False)
+                fs8 = self.provider.get_fsigma8(self.z)
+                s8 = self.provider.get_sigma8_z(self.z)
+                growth = np.mean(Pk_lin/Pk_lin[0], axis = -1)
+                fgrowth = fs8/s8
+                growth = np.flip(growth)
+                fgrowth = np.flip(fgrowth)
+
+                # Undo h units
+                #k = kh*h
+                #Pk_lin = Pk_lin/h**3.
 
                 # np.flip(arr, axis=0) flips the rows of arr, thus making Pk with z
                 # in descending order.
                 Pk_lin = np.flip(Pk_lin, axis=0)
-                cosmo._set_linear_power_from_arrays(a, k, Pk_lin)
+                cosmo._init_pklin({'a': a,
+                        'k': k,
+                        'delta_matter:delta_matter': Pk_lin})
+                cosmo._init_growth({'a': a,
+                        'growth_factor': growth,
+                        'growth_rate': fgrowth})
```

Comment:
Is this kosher CCL API usage?  My understanding from @damonge is that growth and pk should get set as part of the initialization of the `CosmologyCalculator`.  Maybe we can move the `cosmo` initialization happen later on after the growth, pk stuff is figured out?

Reply:
Yes, indeed. This should be done when initializing the `CosmologyCalculator`. Calling any `_whatever` method is a bad sign :-). Incidentally, you no longer need to pass growth arrays to CCL if you don't want to (no core quantities depend on them anymore).

Reply:
Ah that's great, I will change that, thanks!